### PR TITLE
Extreme tests 21 and 27 decimals

### PR DESCRIPTION
**Pull request by julianmrodri**

### CODE REVIEW

**Comment by tbrent** [comment]
Someone else should review this PR since I did the implementation changes.

Roughly: went through all `shiftl_toUint()` calls and added specific rounding. This was almost entirely FLOOR but ended up being CEIL in one or two places. Also there's new math in GnosisTrade to check.

**Review by akshatmittal — APPROVED** [review]
Approving with the caveat that we'll do at least a couple audits on this codebase, with special considerations for this.

**Pull request merged**
